File: src/prism/sarif.cpp
// SARIF 2.1.0 export and CI exit policy. Port of the Python engine
// prism/sarif.py; tests/test_sarif.py compares the two on the same report.

#include "prism/pipeline.hpp"
#include "prism/laws.hpp"

#include <nlohmann/json.hpp>

#include <algorithm>
#include <cctype>
#include <fstream>
#include <map>
#include <string>
#include <system_error>

namespace prism {
namespace fs = std::filesystem;
using ojson = nlohmann::ordered_json;

namespace {

bool is_defect(const std::string& s) {
    return s == laws::FAILED || s == laws::CRASH || s == laws::SANFAIL;
}

bool is_gap(const std::string& s) {
    return s == laws::NOTRUN || s == laws::ERROR || s == laws::TIMEOUT;
}

std::string lower(std::string s) {
    for (char& c : s) c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    return s;
}

std::optional<std::string> level_of(const Finding& f) {
    if (is_defect(f.status)) {
        std::string sev;
        if (auto it = f.extra.find("severity"); it != f.extra.end()) sev = lower(it->second);
        if (sev == "warning" || sev == "style" || f.strength == laws::STRENGTH_SOME)
            return "warning";
        return "error";
    }
    if (f.status == laws::HYPOTHESIS) return "note";
    return std::nullopt;
}

std::string uri_of(std::string p) {
    std::string o;
    for (char c : p) {
        if (c == '\\') o += '/';
        else if (c == '%') o += "%25";
        else if (c == ' ') o += "%20";
        else o += c;
    }
    return o;
}

// pathlib.Path.as_uri(): percent-encode everything but unreserved and '/'.
std::string file_uri(const fs::path& p) {
    static const char* hex = "0123456789ABCDEF";
    std::string g = p.generic_string();
    if (g.empty() || g.front() != '/') g = "/" + g;
    std::string o = "file://";
    for (unsigned char c : g) {
        if (std::isalnum(c) || c == '-' || c == '.' || c == '_' || c == '~' || c == '/') {
            o += static_cast<char>(c);
        } else {
            o += '%';
            o += hex[c >> 4];
            o += hex[c & 15];
        }
    }
    return o;
}

std::string rstrip_colon_space(std::string s) {
    while (!s.empty() && (s.back() == ':' || s.back() == ' ')) s.pop_back();
    return s;
}

}  // namespace

std::string to_sarif(const RunReport& report) {
    std::map<std::string, ojson> rules;
    ojson results = ojson::array();
    for (auto& s : report.stages) {
        for (auto& f : s.findings) {
            auto level = level_of(f);
            if (!level) continue;
            std::string rid = f.cls.empty() ? f.stage + "/" + f.status : f.cls;
            if (!rules.contains(rid))
                rules[rid] = ojson{{"id", rid}, {"shortDescription", {{"text", rid}}}};
            std::string text = "[" + f.stage + " " + f.status + "] " + f.message;
            if (!f.counterexample.empty()) text += " (counterexample: " + f.counterexample + ")";
            ojson res = {
                {"ruleId", rid},
                {"level", *level},
                {"message", {{"text", text}}},
                {"properties", {{"stage", f.stage}, {"status", f.status}, {"strength", f.strength}}},
            };
            if (f.function && !f.function->empty()) res["properties"]["function"] = *f.function;
            if (f.status == laws::HYPOTHESIS) res["properties"]["hypothesis"] = true;
            if (!f.file.empty()) {
                ojson phys = {{"artifactLocation", {{"uri", uri_of(f.file)}, {"uriBaseId", "SRCROOT"}}}};
                if (f.line && *f.line > 0) phys["region"] = {{"startLine", *f.line}};
                res["locations"] = ojson::array({{{"physicalLocation", phys}}});
            }
            results.push_back(std::move(res));
        }
    }
    ojson notes = ojson::array();
    bool crashed = false;
    for (auto& s : report.stages) {
        if (s.status == "failed") crashed = true;
        if (s.status != "NOTRUN" && s.status != "failed") continue;
        notes.push_back({{"level", s.status == "failed" ? "error" : "warning"},
                         {"message",
                          {{"text", rstrip_colon_space(s.name + " " + s.status + ": " +
                                                       (s.detail.empty() ? s.install : s.detail))}}}});
    }
    ojson rule_arr = ojson::array();
    for (auto& [id, r] : rules) rule_arr.push_back(r);  // std::map: sorted by id
    ojson run = {
        {"tool",
         {{"driver",
           {{"name", "PRISM"},
            {"version", PRISM_VERSION},
            {"informationUri", "https://github.com/odin-loki/PRISM"},
            {"rules", rule_arr}}}}},
        {"invocations",
         ojson::array({{{"executionSuccessful", !crashed}, {"toolExecutionNotifications", notes}}})},
        {"results", results},
        {"properties", {{"confidence", report.confidence}}},
    };
    if (!report.root.empty()) {
        std::error_code ec;
        fs::path root(report.root);
        fs::path base = fs::is_directory(root, ec) ? root : root.parent_path();
        auto canon = fs::weakly_canonical(base, ec);
        run["originalUriBaseIds"] = {{"SRCROOT", {{"uri", file_uri(ec ? base : canon) + "/"}}}};
    }
    ojson doc = {{"$schema", "https://json.schemastore.org/sarif-2.1.0.json"},
                 {"version", "2.1.0"},
                 {"runs", ojson::array({run})}};
    return doc.dump(2);
}

void write_sarif(const RunReport& report, const fs::path& path) {
    std::error_code ec;
    fs::create_directories(path.parent_path(), ec);
    std::ofstream(path, std::ios::binary) << to_sarif(report);
}

int exit_code(const RunReport& report, std::string_view fail_on) {
    for (auto& s : report.stages)
        if (s.status == "failed") return 2;
    if (fail_on == "never") return 0;
    for (auto& s : report.stages)
        for (auto& f : s.findings)
            if (is_defect(f.status)) return 1;
    if (fail_on == "gap") {
        for (auto& s : report.stages) {
            if (s.status == "NOTRUN") return 1;
            for (auto& f : s.findings)
                if (is_gap(f.status)) return 1;
        }
    }
    return 0;
}

}  // namespace prism

File: prism/sarif.py
"""SARIF 2.1.0 export and CI exit policy. Same output as src/prism/sarif.cpp.

SARIF is what GitHub code scanning, VS Code and most CI dashboards read.
Laws still hold in this format:
  - only defects become results (FAILED / CRASH / SANFAIL); LLM output is a
    `note`, never an error, and is marked as a hypothesis;
  - a stage that could not run is a tool execution notification, so a
    missing tool is visible, not silently absent;
  - nothing here claims a proof.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from prism import __version__, laws
from prism.models import Finding, RunReport

SARIF_SCHEMA = "https://json.schemastore.org/sarif-2.1.0.json"
INFO_URI = "https://github.com/odin-loki/PRISM"

DEFECT_STATUSES = (laws.FAILED, laws.CRASH, laws.SANFAIL)
GAP_STATUSES = (laws.NOTRUN, laws.ERROR, laws.TIMEOUT)
FAIL_ON = ("never", "defect", "gap")
# --fail-on counts a defect only at error level: a finding whose
# extra.severity is one of these (a compiler/linter warning) is reported
# (SARIF level "warning") but does not fail the build.
NON_BLOCKING_SEVERITIES = ("warning", "note", "style")


def blocks(f: Finding) -> bool:
    """A defect --fail-on defect/gap counts (not a warning/note/style)."""
    if f.status not in DEFECT_STATUSES:
        return False
    return str((f.extra or {}).get("severity", "")).lower() not in NON_BLOCKING_SEVERITIES


def _level(f: Finding) -> str | None:
    if f.status in DEFECT_STATUSES:
        sev = str((f.extra or {}).get("severity", "")).lower()
        if sev in {"warning", "style"} or f.strength == laws.STRENGTH_SOME:
            return "warning"
        return "error"
    if f.status == laws.HYPOTHESIS:
        return "note"
    return None


def _uri(path: str) -> str:
    return path.replace("\\", "/").replace("%", "%25").replace(" ", "%20")


def _rule_id(f: Finding) -> str:
    return f.cls or f"{f.stage}/{f.status}"


def to_sarif(report: RunReport) -> dict[str, Any]:
    rules: dict[str, dict[str, Any]] = {}
    results: list[dict[str, Any]] = []
    for s in report.stages:
        for f in s.findings:
            level = _level(f)
            if level is None:
                continue
            rid = _rule_id(f)
            if rid not in rules:
                rules[rid] = {"id": rid, "shortDescription": {"text": rid}}
            text = f"[{f.stage} {f.status}] {f.message}"
            if f.counterexample:
                text += f" (counterexample: {f.counterexample})"
            res: dict[str, Any] = {
                "ruleId": rid,
                "level": level,
                "message": {"text": text},
                "properties": {
                    "stage": f.stage,
                    "status": f.status,
                    "strength": f.strength,
                },
            }
            if f.function:
                res["properties"]["function"] = f.function
            if f.status == laws.HYPOTHESIS:
                res["properties"]["hypothesis"] = True
            if f.file:
                phys: dict[str, Any] = {
                    "artifactLocation": {"uri": _uri(f.file), "uriBaseId": "SRCROOT"},
                }
                if f.line and int(f.line) > 0:
                    phys["region"] = {"startLine": int(f.line)}
                res["locations"] = [{"physicalLocation": phys}]
            results.append(res)
    notes = [
        {
            "level": "error" if s.status == "failed" else "warning",
            "message": {"text": f"{s.name} {s.status}: {s.detail or s.install}".rstrip(": ")},
        }
        for s in report.stages if s.status in {"NOTRUN", "failed"}
    ]
    root = Path(report.root) if report.root else None
    base = root if root is None or root.is_dir() else root.parent
    run: dict[str, Any] = {
        "tool": {"driver": {
            "name": "PRISM",
            "version": __version__,
            "informationUri": INFO_URI,
            "rules": sorted(rules.values(), key=lambda r: r["id"]),
        }},
        "invocations": [{
            "executionSuccessful": not any(s.status == "failed" for s in report.stages),
            "toolExecutionNotifications": notes,
        }],
        "results": results,
        "properties": {"confidence": report.confidence},
    }
    if base is not None:
        run["originalUriBaseIds"] = {"SRCROOT": {"uri": base.resolve().as_uri() + "/"}}
    return {"$schema": SARIF_SCHEMA, "version": "2.1.0", "runs": [run]}


def write_sarif(report: RunReport, path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(to_sarif(report), indent=2), encoding="utf-8")


def exit_code(report: RunReport, fail_on: str) -> int:
    """2 = a stage crashed; 1 = --fail-on policy tripped; 0 otherwise.

    Defects with extra.severity warning/note/style never trip the policy.
    """
    if any(s.status == "failed" for s in report.stages):
        return 2
    if fail_on == "never":
        return 0
    findings = [f for s in report.stages for f in s.findings]
    if any(blocks(f) for f in findings):
        return 1
    if fail_on == "gap" and (
        any(s.status == "NOTRUN" for s in report.stages)
        or any(f.status in GAP_STATUSES for f in findings)
    ):
        return 1
    return 0
